// fastify-metrics/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
const fastify_plugin_1 = __importDefault(require("fastify-plugin"));
const prom_client_1 = __importDefault(require("prom-client"));
/**
 * Fastify metrics plugin
 */
const fastifyMetricsPlugin = function fastifyMetrics(fastify, { enableDefaultMetrics = true, groupStatusCodes = false, pluginName = 'metrics', blacklist, register, prefix, endpoint, metrics = {} } = {}, next) {
    const plugin = { client: prom_client_1.default };
    if (enableDefaultMetrics) {
        const collectMetricsForUrl = (url) => {
            const queryIndex = url.indexOf('?');
            url = queryIndex === -1 ? url : url.substring(0, queryIndex);
            if (!blacklist) {
                return true;
            }
            if (Array.isArray(blacklist)) {
                return blacklist.indexOf(url) === -1;
            }
            if (typeof blacklist === 'string') {
                return blacklist !== url;
            }
            if (typeof blacklist.test === 'function') {
                return !blacklist.test(url);
            }
            return false;
        };
        const defaultOpts = {};
        const opts = {
            histogram: {
                name: 'http_request_duration_seconds',
                help: 'request duration in seconds',
                labelNames: ['status_code', 'method', 'route'],
                buckets: [0.05, 0.1, 0.5, 1, 3, 5, 10]
            },
            summary: {
                name: 'http_request_summary_seconds',
                help: 'request duration in seconds summary',
                labelNames: ['status_code', 'method', 'route'],
                percentiles: [0.5, 0.9, 0.95, 0.99]
            }
        };
        if (register) {
            plugin.clearRegister = register.clear;
            defaultOpts.register = register;
            opts.histogram.registers = [register];
            opts.summary.registers = [register];
        }
        if (prefix) {
            defaultOpts.prefix = prefix;
            opts.histogram.name = `${prefix}${opts.histogram.name}`;
            opts.summary.name = `${prefix}${opts.summary.name}`;
        }
        Object.keys(metrics)
            .filter(opts.hasOwnProperty.bind(opts))
            .forEach(key => {
            Object.assign(opts[key], metrics[key]);
        });
        prom_client_1.default.collectDefaultMetrics(defaultOpts);
        const routeHist = new prom_client_1.default.Histogram(opts.histogram);
        const routeSum = new prom_client_1.default.Summary(opts.summary);
        if (endpoint) {
            fastify.route({
                url: endpoint,
                method: 'GET',
                schema: { hide: true },
                handler: (_, reply) => {
                    const data = register
                        ? register.metrics()
                        : prom_client_1.default.register.metrics();
                    reply.type('text/plain').send(data);
                }
            });
        }
        fastify.addHook('onRequest', (request, _, next) => {
            if (request.req.url && collectMetricsForUrl(request.req.url)) {
                request.metrics = {
                    hist: routeHist.startTimer(),
                    sum: routeSum.startTimer()
                };
            }
            next();
        });
        fastify.addHook('onResponse', function (request, reply, next) {
            if (request.metrics) {
                let routeId = reply.context.config.url || request.req.url;
                if (reply.context.config.statsId) {
                    routeId = reply.context.config.statsId;
                }
                const method = request.req.method;
                const statusCode = groupStatusCodes
                    ? `${Math.floor(reply.res.statusCode / 100)}xx`
                    : reply.res.statusCode;
                request.metrics.sum({
                    method: method || 'UNKNOWN',
                    route: routeId,
                    status_code: statusCode
                });
                request.metrics.hist({
                    method: method || 'UNKNOWN',
                    route: routeId,
                    status_code: statusCode
                });
            }
            next();
        });
    }
    fastify.decorate(pluginName, plugin);
    next();
};
module.exports = fastify_plugin_1.default(fastifyMetricsPlugin, {
    fastify: '>=2.0.0',
    name: 'fastify-metrics'
});
